clash-lib/src/Clash/Netlist/Expr.hs

```diff
@@ -0,0 +1,166 @@
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Clash.Netlist.Expr where
+
+import Control.Monad (zipWithM)
+import Control.Exception (assert)
+import Data.Bits (Bits, testBit, setBit, zeroBits)
+import Data.Foldable (fold)
+import Data.Tree (Tree(..))
+import GHC.Stack (HasCallStack)
+import Clash.Primitives.DSL (tySize)
+import Clash.Netlist.Types
+  ( Size, Bit(..), Expr(..), HWType(..), Literal(..), Modifier(..)
+  , BlackBoxContext(..)
+  )
+
+-- | Turns a constant expression of known bitsize into their
+-- corresponding bitstream representation, arranged as a tree
+-- that corresponds to the structure of the expression.
+--
+-- NOTE: This conversion serves as a best effort approach and can be
+-- considered a hack. Fully featured constant expression evaluation is
+-- not available in clash yet and will replace this implementation
+-- once it is officially supported.
+bits :: HasCallStack => Size -> Expr -> Either Expr (Tree [Bool])
+bits size expr = case expr of
+  Literal _ lit -> case lit of
+    BitLit bLit   -> case bLit of
+      H -> leaf [True]
+      L -> leaf [False]
+      _ -> Left expr
+    BoolLit bLit      -> leaf [bLit]
+    NumLit nLit       -> leaf $ toBits size nLit
+    BitVecLit _ bvLit -> leaf $ toBits size bvLit
+    VecLit lits       ->
+      mapM (bits (size `div` length lits) . lit2Expr) lits >>= inner
+    StringLit{}       -> Left expr
+  DataCon ty m subs -> assert (tySize ty == size) $ case ty of
+    Vector s t      -> vecBits (tySize t) s subs
+    Product _ _ tys -> zipWithM bits (map tySize tys) subs >>= inner
+    Sum _ cs        -> spBits expr size m subs $ map (const []) cs
+    SP _ xs         -> spBits expr size m subs $ map (map tySize . snd) xs
+    _               -> case subs of
+      [e] -> bits size e
+      []  -> leaf []
+      _   -> Left expr
+  -- appears in case of complex transformations, e.g.,
+  -- >>> (bv2v 0b010) :: Vec 3 Bit
+  -- >>> (iterate (SNat @3) not True) :: Vec 3 Bool
+  -- >>> (complement <$> (True :> False :> Nil)) :: Vec 2 Bool
+  Identifier{} -> Left expr
+  DataTag{} -> Left expr
+  BlackBoxE bbName _ _ _ _ bbCtx _ -> case bbName of
+    "Clash.Sized.Internal.BitVector.low" -> leaf [False]
+    "Clash.Sized.Internal.BitVector.high" -> leaf [True]
+    "Clash.Sized.Vector.replicate" -> case bbInputs bbCtx of
+      [ (eSize, ty, _), (eValue, _, _) ] -> do
+        bs <- bits (tySize ty) eSize
+        let s = fromBits $ fold bs
+        v <- bits (size `div` s) eValue
+        inner $ replicate s v
+      _ -> Left expr
+    _ ->
+      if bbName `elem` skippableBBs
+      then skippableBBBits expr bbCtx size
+      else Left expr
+  ToBv _ _ e -> bits size e
+  FromBv _ _ e -> bits size e
+  IfThenElse cond match alt -> case bits 1 cond of
+    Right (Node [True] [])  -> bits size match
+    Right (Node [False] []) -> bits size alt
+    _ -> Left expr
+  Noop -> leaf []
+
+ where
+  -- known skippable blackboxes
+  skippableBBs =
```

Comment:
```suggestion
  skippableBBs = Set.fromList
```

I'm under no illusion that this will have tangible performance gains right now, but let's keep the computational complexity low in case this list grows.